sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -289,13 +294,13 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   /**
    * Returns a copy of this node where `f` has been applied to all the nodes children.
    */
-  def mapChildren(f: BaseType => BaseType): BaseType = {
-    if (children.nonEmpty) {
+  def mapChildren(f: BaseType => BaseType, forceCopy: Boolean = false): BaseType = {
```

Comment:
do we really need this? Can we avoid modifying this method?

Reply:
Well, I need this for `makeDeepCopy()` and didn't want to copy the logic. Any suggestion?

Reply:
I see, my point is that this is a very critical method, so I'd prefer to leave it unchanged. Do we really need that makeDeepCopy? If so, why do we need it? I hope we can rid of it somehow...

Reply:
I see. I run into an issue with `BroadcastExchangeExec` which can't be reused because of its `relationFuture`.

Reply:
ok, now I see the issue here. But I think this is not the best way to go for that. I think this is the same reason why you are adding some exclusions in the prepare method (please correct me if I am wrong). I'd rather introduce a `clean` method in `SparkPlan` which by default does nothing, but in case of broadcasts cleans the broadcast and hence force recomputation at the next step. What do you think? cc @cloud-fan @dongjoon-hyun  @gatorsmile fot their opinion on this too

Reply:
You are right @mgaido91, `BroadcastExchangeExec` and similar nodes cause some issues there. We can't prepare them until we have the previous iteration results (or the anchor) This is why I introduced `doNotPrepareInAdvance`.

Reply:
@mgaido I reverted the changes in TreeNode and introduced `reset()` on `SparkPlan` as you suggested. Also fixed the case when we have an `ExchangeCoordinator` and added test for it.